v2/api/appconfiguration/customizations/configuration_store_extensions.go

```diff
@@ -0,0 +1,148 @@
+	"k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/secrets"
+)
+
+var _ extensions.SecretsRetriever = &ConfigurationStoreExtension{}
+
+func (ext *ConfigurationStoreExtension) RetrieveSecrets(
+	ctx context.Context,
+	obj genruntime.ARMMetaObject,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger) ([]*v1.Secret, error) {
+
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	typedObj, ok := obj.(*storage.ConfigurationStore)
+	if !ok {
+		return nil, errors.Errorf("cannot run on unknown resource type %T, expected *appconfiguration.ConfigurationStore", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = typedObj
+
+	hasSecrets := secretsSpecified(typedObj)
+	if !hasSecrets {
+		log.V(Debug).Info("No secrets retrieval to perform as operatorSpec is empty")
+		return nil, nil
+	}
+
+	id, err := genruntime.GetAndParseResourceID(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]armappconfiguration.APIKey)
+	// Only bother calling ListKeys if there are secrets to retrieve
+	if hasSecrets {
+		subscription := armClient.SubscriptionID()
+		// Using armClient.ClientOptions() here ensures we share the same HTTP connection, so this is not opening a new
+		// connection each time through
+		var confClient *armappconfiguration.ConfigurationStoresClient
+		confClient, err = armappconfiguration.NewConfigurationStoresClient(subscription, armClient.Creds(), armClient.ClientOptions())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create new ConfigurationStoresClient")
+		}
+
+		var pager *runtime.Pager[armappconfiguration.ConfigurationStoresClientListKeysResponse]
+		var resp armappconfiguration.ConfigurationStoresClientListKeysResponse
+		pager = confClient.NewListKeysPager(id.ResourceGroupName, obj.AzureName(), nil)
+		for pager.More() {
+			resp, err = pager.NextPage(ctx)
+			secretsByName(resp.Value, keys)
+
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to retreive response")
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed listing keys")
+		}
+
+	}
+
+	secretSlice, err := secretsToWrite(typedObj, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	return secretSlice, nil
+}
+
+func secretsSpecified(obj *storage.ConfigurationStore) bool {
+	if obj.Spec.OperatorSpec == nil || obj.Spec.OperatorSpec.Secrets == nil {
+		return false
+	}
+
+	secrets := obj.Spec.OperatorSpec.Secrets
+
+	if secrets.PrimaryKeyID != nil ||
```

Comment:
I actually think you can remove `PrimaryKeyID` and all of the other `ID` fields? From what I can tell, it's just a readonly ID that identifies if the key has changed. I don't think their clients actually use/need it, it's like the `LastModifiedTime`. There in the response but doesn't need to be exposed to the pods?

Reply:
I had a same thought on IDs, Then I ended up creating a resource in Azure and checked on portal. While reading the Access Keys, there it shows ID to be `credential`. That's the only reason for adding Ids
![image](https://user-images.githubusercontent.com/38904804/188043716-ebb08a86-09a5-4825-9a30-f17f3a16834e.png)

Reply:
Its the same for Read-Write keys btw.

Reply:
I think thats their UI weirdness. The example API response shows :
```
    {
      "id": "439AD01B4BE67DB1",
      "name": "Primary",
      "value": "000000000000000000000000000000000000000000000000000000",
      "connectionString": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "lastModified": "2018-04-24T16:30:54+00:00",
      "readOnly": false
    },
```

The `ID` field isnt long enough to be what they are showing as the `credential`.

Reply:
I think the example API is not correct or too old? Check the response below.

List keys using `az rest` on resource:

![image](https://user-images.githubusercontent.com/38904804/188297179-1cf0fa11-2066-476e-be0c-e0d18e9c39a3.png)

On portal: 

![image](https://user-images.githubusercontent.com/38904804/188297201-1449dde6-27be-4618-a9c7-04035b50848b.png)

Reply:
Hmm, the fact that it's in the connection string suggests it may actually be needed, although I can't find any examples of theirs showing its use (everybody uses the connectionString).

Alright, happy to keep it then.